src/Compilers/Core/Portable/Compilation.EmitStream.cs

```diff
@@ -0,0 +1,176 @@
+            {
+                _tempInfo?.tempStream.Dispose();
+                _tempInfo = null;
+
+                // The _stream value is deliberately excluded from being disposed here. That value is not 
+                // owned by this type.
+                _stream = null;
+            }
+
+            /// <summary>
+            /// Create the stream which should be used for Emit. This should only be called one time.
+            /// </summary>
+            private Stream CreateStream(DiagnosticBag diagnostics)
+            {
+                Debug.Assert(_stream == null);
+                Debug.Assert(diagnostics != null);
+
+                if (diagnostics.HasAnyErrors())
+                {
+                    return null;
+                }
+
+                _stream = _emitStreamProvider.GetOrCreateStream(diagnostics);
+                if (_stream == null)
+                {
+                    Debug.Assert(diagnostics.HasAnyErrors());
+                    return null;
+                }
+
+                // If the current strong name provider is the Desktop version, signing can only be done to on-disk files.
+                // If this binary is configured to be signed, create a temp file, output to that
+                // then stream that to the stream that this method was called with. Otherwise output to the
+                // stream that this method was called with.
+                if (_emitStreamSignKind == EmitStreamSignKind.SignedWithFile)
+                {
+                    Debug.Assert(_strongNameProvider != null);
+
+                    try
+                    {
+                        var fileSystem = _strongNameProvider.FileSystem;
+                        Func<string, Stream> streamConstructor = path => fileSystem.CreateFileStream(path, FileMode.Create, FileAccess.ReadWrite, FileShare.ReadWrite);
+
+                        var tempDir = fileSystem.GetTempPath();
+                        var tempFilePath = Path.Combine(tempDir, Guid.NewGuid().ToString("N"));
+                        var tempStream = FileUtilities.CreateFileStreamChecked(streamConstructor, tempFilePath);
+
+                        _tempInfo = (tempStream, tempFilePath);
+                        return tempStream;
+                    }
+                    catch (IOException e)
+                    {
+                        throw new Cci.PeWritingException(e);
+                    }
+                }
+                else
+                {
+                    return _stream;
+                }
+            }
+
+            internal bool Complete(StrongNameKeys strongNameKeys, CommonMessageProvider messageProvider, DiagnosticBag diagnostics)
+            {
+                Debug.Assert(_stream != null);
+
+                if (_tempInfo.HasValue)
+                {
+                    Debug.Assert(_emitStreamSignKind == EmitStreamSignKind.SignedWithFile);
+                    var (tempStream, tempFilePath) = _tempInfo.Value;
```

Comment:
>Value [](start = 63, length = 5)

```GetValueOrDefault()```?

Reply:
There is a check for `HasValue` just above hence the `OrDefault` shouldn't be necessary here. Am I missing something?

---
In reply to: [245390363](https://github.com/dotnet/roslyn/pull/31986#discussion_r245390363) [](ancestors = 245390363)

Reply:
> There is a check for HasValue just above hence the OrDefault shouldn't be necessary here. Am I missing something?

It is a minor thing, ```GetValueOrDefault``` avoids additional null check

---
In reply to: [245392007](https://github.com/dotnet/roslyn/pull/31986#discussion_r245392007) [](ancestors = 245392007,245390363)

Reply:
Ah okay.

---
In reply to: [245393074](https://github.com/dotnet/roslyn/pull/31986#discussion_r245393074) [](ancestors = 245393074,245392007,245390363)